common/quotas/clocked_rate_limiter.go

```diff
@@ -137,6 +139,8 @@ func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
 		return fmt.Errorf("%w: %v", ErrRateLimiterWaitInterrupted, ctx.Err())
 	case <-waitExpired:
 		return nil
+	case <-l.recycleCh:
+		return nil
```

Comment:
There could be waiters waiting for more than one token. But this will unblock them as well. Ideally RecycleToken() should only recycle 1 token. If it is not straightforward to track token count, and is not needed for this use case, we should at least add a warning in the RecycleToken() comment.

Reply:
The matching code where we need to use `RecycleToken` only ever reserves one token (same with most of our rate limiter uses -- calling `WaitN` directly is rare), so I decided to make this safe by only letting the recycle go through if the waiter is waiting on one token. If a waiter is waiting on >1 token, they will not be unblocked.

Added the following to the RecycleToken() comment:
```
// Since we don't know how many tokens were reserved by the process calling recycle, we will only unblock
// new reservations that are for one token (otherwise we could recycle a 1-token-reservation and unblock
// a 100-token-reservation). If all waiting processes are waiting for >1 tokens, this is a no-op.
```

Reply:
One reason why it's not straightforward to track token count is because we don't explicitly know how many tokens were reserved when we call recycle.

In the case of matching at least, we do know that it's always 1, because we only ever take 1 token, so the caller of `RecycleToken()` could pass 1 to the recycle call since it knows from the code that it only took 1 token previously. That could definitely be abused though.

Another solution I considered was to add an atomic uint to the rate limiter struct, and increment it with whatever `n` `RecycleToken()` is called with. Then, when the `recycleCh` receives a signal that a recycle just occurred, the waiter checks if there are enough tokens in the bucket to satisfy the reservation, and if there are, it decrements the atomic and takes the reservation, but if there are not, it does nothing and keeps waiting in the select.

I think because most of our RL reservations use only 1 token, the method I pushed a few minutes ago is simpler to test and reason about, but I'm open to the bucket method if you think that it's better for some reason.

Reply:
Thanks! I think that is good enough for now. We can add code to handle multiple tokens when we have a use case for that.

Yeah, I was also thinking how to handle multiple tokens. I was thinking if it is better to add `Wait()` and `Recycle()` methods to `Reservation` interface. `Reservation` object will know the number of tokens to recycle. That way we don't need to pass the token count to `Recycle()`. 

One idea is to pass the number of tokens to recycle to `l.recycleCh` instead of empty struct. Then the receiving waiter could check if that is enough to unblock itself. If not, it could pass the same number back to the channel for other waiters. I am not suggesting to implement any of these. It is fine to implement it when we have a use case for that.

Reply:
> I was thinking if it is better to add Wait() and Recycle() methods to Reservation interface. Reservation object will know the number of tokens to recycle. That way we don't need to pass the token count to Recycle().
The problem with that idea is that `Wait` doesn't actually return the reservation to the caller, it just takes the reservation and then returns error or nil. So, the caller of `Wait`, which is the process that knows when to call `Recycle`, doesn't have access to the reservation that it waited on.

Reply:
I meant the caller to use Reserve() method first to get a reservation and then call Wait() on the reservation instead of calling Wait() on the ratelimiter.

Reply:
I agree that this is overkill for now. We should just comment on RecycleToken that it recycles a single token and is only useful for the case where you took one.